File: code/python/ch04/learn/example_01_basic_functions.py
"""
Example 01: Basic Functions
===========================
Chapter 4: Functions

Run with:
    python code/python/ch04/learn/example_01_basic_functions.py
"""

# ============================================================
# PART 1: Defining a simple function
# ============================================================
# A function is a reusable block of code with a name.
# You define it with `def`, give it a name, and optionally list parameters.

def square(x):
    """Return the square of x."""
    return x * x


def greet(name):
    """Print a greeting for the given name."""
    print(f"Hey there, {name}! Welcome aboard.")


print("=== PART 1: Simple Functions ===")
print(f"square(5) = {square(5)}")      # 25
print(f"square(-3) = {square(-3)}")    # 9
greet("Alex")                           # Hey there, Alex! Welcome aboard.
print()

# ============================================================
# PART 2: Functions with multiple parameters
# ============================================================
# Functions can take more than one parameter, separated by commas.

def add(a, b):
    """Return the sum of a and b."""
    return a + b


def format_name(first, last):
    """Return a nicely formatted full name."""
    return f"{first} {last}"


print("=== PART 2: Multiple Parameters ===")
print(f"add(3, 7) = {add(3, 7)}")                         # 10
print(f"add(-1, 1) = {add(-1, 1)}")                        # 0
print(f"format_name('Ada', 'Lovelace') = {format_name('Ada', 'Lovelace')}")  # Ada Lovelace
print()

# ============================================================
# PART 3: Functions with return values vs. print
# ============================================================
# A function that uses `return` gives back a value you can store or use.
# A function that only uses `print` shows something on screen but returns None.

def multiply_return(a, b):
    """Return a * b — the caller decides what to do with the result."""
    return a * b


def multiply_print(a, b):
    """Print a * b — no return value (returns None implicitly)."""
    print(f"{a} * {b} = {a * b}")


print("=== PART 3: Return vs. Print ===")
result = multiply_return(4, 5)
print(f"multiply_return(4, 5) returned: {result}")  # 20
print(f"Can we use it in math? result + 10 = {result + 10}")  # 30
print()

print("multiply_print(4, 5) output:")
none_result = multiply_print(4, 5)   # prints "4 * 5 = 20"
print(f"But its return value is: {none_result}")  # None
# print(none_result + 10)  # This would crash! Can't add None + 10
print()

# ============================================================
# PART 4: Default parameters
# ============================================================
# You can give parameters default values. If the caller doesn't
# provide that argument, the default kicks in.

def repeat_string(s, n=3):
    """Return the string s repeated n times, separated by spaces."""
    return " ".join([s] * n)


print("=== PART 4: Default Parameters ===")
print(f"repeat_string('ha') = '{repeat_string('ha')}'")          # ha ha ha  (n defaults to 3)
print(f"repeat_string('yo', 5) = '{repeat_string('yo', 5)}'")    # yo yo yo yo yo
print(f"repeat_string('!', 1) = '{repeat_string('!', 1)}'")      # !
print(f"repeat_string('x', 0) = '{repeat_string('x', 0)}'")      # (empty string)
print()

# ============================================================
# PART 5: Calling one function from another
# ============================================================
# Functions can call other functions. This is how you build
# complex behavior from simple building blocks.

def min_of_two(a, b):
    """Return the smaller of a and b."""
    if a <= b:
        return a
    return b


def min_of_three(a, b, c):
    """Return the smallest of a, b, and c using min_of_two."""
    return min_of_two(min_of_two(a, b), c)


print("=== PART 5: Functions Calling Functions ===")
print(f"min_of_two(7, 3) = {min_of_two(7, 3)}")          # 3
print(f"min_of_two(-1, -5) = {min_of_two(-1, -5)}")      # -5
print(f"min_of_three(9, 2, 7) = {min_of_three(9, 2, 7)}")  # 2
print(f"min_of_three(4, 4, 4) = {min_of_three(4, 4, 4)}")  # 4
print(f"min_of_three(-1, 0, 1) = {min_of_three(-1, 0, 1)}")  # -1

File: code/java/ch04/learn/Example01BasicFunctions.java
package ch04.learn;

/**
 * Example 01: Basic Functions
 * ===========================
 * Chapter 4: Functions
 *
 * This file shows you how Java defines and uses methods (functions).
 * You'll see basic method definitions, multiple parameters, return values,
 * method overloading, and functions calling functions.
 * Read through each section and run the file to see the output.
 *
 * Build and run:
 *   cd code/java
 *   javac ch04/learn/Example01BasicFunctions.java
 *   java ch04.learn.Example01BasicFunctions
 */
public class Example01BasicFunctions {

    // ── 1. Basic Method Definition ───────────────────────────────────
    // A method has: access modifier, return type, name, parameters, body.
    // 'static' means we can call it without creating an object.

    /** Prints a friendly greeting. No return value (void). */
    static void greet() {
        System.out.println("Hello from a method!");
    }

    // ── 2. Parameters and Return Values ──────────────────────────────
    // Parameters are the inputs; return type says what comes back.

    /** Returns the square of a number. */
    static int square(int n) {
        return n * n;
    }

    /** Adds two numbers and returns the result. */
    static int add(int a, int b) {
        return a + b;
    }

    // ── 3. Multiple Parameters and String Return ─────────────────────

    /** Builds a full name from first and last. */
    static String fullName(String first, String last) {
        return first + " " + last;
    }

    /** Checks if a number is in range [low, high] inclusive. */
    static boolean inRange(int value, int low, int high) {
        return value >= low && value <= high;
    }

    // ── 4. Method Overloading (Java's Way to Handle "Default" Params) ─
    // In Python, you can write def greet(name="World").
    // Java doesn't have default parameters, but you can define
    // multiple methods with the same name but different parameter lists.

    /** Greet a specific person. */
    static String greetPerson(String name) {
        return "Hello, " + name + "!";
    }

    /** Greet with a custom greeting. */
    static String greetPerson(String greeting, String name) {
        return greeting + ", " + name + "!";
    }

    /** Greet with greeting, name, and punctuation. */
    static String greetPerson(String greeting, String name, String punctuation) {
        return greeting + ", " + name + punctuation;
    }

    // ── 5. Functions Calling Functions ────────────────────────────────
    // Methods can call other methods. This is how you build complex
    // programs from simple building blocks.

    /** Calculate area of a rectangle. */
    static int rectangleArea(int width, int height) {
        return width * height;
    }

    /** Calculate area of a square (calls rectangleArea). */
    static int squareArea(int side) {
        return rectangleArea(side, side);   // reuse!
    }

    /** Calculate volume of a box (calls rectangleArea for the base). */
    static int boxVolume(int length, int width, int height) {
        int baseArea = rectangleArea(length, width);
        return baseArea * height;
    }

    // ── Main ─────────────────────────────────────────────────────────

    public static void main(String[] args) {

        // 1. Basic method call
        System.out.println("=== 1. Basic Method Definition ===");
        greet();
        System.out.println();

        // 2. Parameters and return values
        System.out.println("=== 2. Parameters and Return Values ===");
        System.out.println("square(5) = " + square(5));
        System.out.println("add(3, 7) = " + add(3, 7));

        // You can use the return value in expressions:
        int result = add(square(3), square(4));
        System.out.println("add(square(3), square(4)) = " + result);
        System.out.println();

        // 3. Multiple parameters
        System.out.println("=== 3. Multiple Parameters ===");
        System.out.println("fullName(\"Ada\", \"Lovelace\") = " + fullName("Ada", "Lovelace"));
        System.out.println("inRange(5, 1, 10) = " + inRange(5, 1, 10));
        System.out.println("inRange(15, 1, 10) = " + inRange(15, 1, 10));
        System.out.println();

        // 4. Method overloading
        System.out.println("=== 4. Method Overloading ===");
        System.out.println(greetPerson("Maya"));
        System.out.println(greetPerson("Hi", "Maya"));
        System.out.println(greetPerson("Hey", "Maya", "!!!"));
        // Java picks the right version based on what you pass in.
        // This is called "overloading" — same name, different signatures.
        System.out.println();

        // 5. Functions calling functions
        System.out.println("=== 5. Functions Calling Functions ===");
        System.out.println("rectangleArea(4, 5) = " + rectangleArea(4, 5));
        System.out.println("squareArea(4) = " + squareArea(4));
        System.out.println("boxVolume(3, 4, 5) = " + boxVolume(3, 4, 5));
    }
}
